Voice Support

LordOfPolls opened the pull request:
Stop asking me when voice; here voice. But it not work good

## What type of pull request is this?

<!-- Check whichever applies to your PR -->
- [ ] Non-breaking code change
- [x] Breaking code change
- [ ] Documentation change/addition
- [ ] Tests change

## Description
<!-- Clearly and concisely describe what this PR is for, and why you feel it should be merged. -->
Adds voice support to the library, by popular request. 

Example:
```python
@dis_snek.slash_command("play", "play a song!")
@dis_snek.slash_option("song", "The song to play", 3, True)
async def play(self, ctx: dis_snek.InteractionContext, song: str):
    if not ctx.voice_state:
        # connect to the authors voice channel
        await ctx.author.voice.channel.connect()    

    # get the requested song with YTDL
    audio = await YTDLAudio.from_url(song)
    await ctx.send(f"Playing **{audio.entry['title']}**")
    await ctx.voice_state.play(audio)  # play
```

## Changes
<!-- - A bullet pointed list outlining the changes you have made -->
- Add required voice components: `voice/`-`audio` `encryption` `opus` `player` `voice_gateway` and `models/snek/active_voice_state`
- Refactor existing gateway to have a parent class with shared methods between `voice_gateway` and `gateway`
- Add helper methods to access a user voice state
- Add helper methods to connect to a vc
- Properly cache guild-creation voice state data
- cache client voice states seperately from the normal voice states
- Refactor build dependecies to go through setup.py
- Add `voice` `all` and `docs` extras 
- Include opus dlls for windows users

## Checklist

<!-- These are actions you **must** have taken, if you haven't, your PR will be rejected -->
- [x] I've formatted my code with [Black](https://black.readthedocs.io/en/stable/)
- [x] I've ensured my code works on `Python 3.10.x`
- [x] I've tested my code

LordOfPolls committed:
feat: unfinished voice support

Stop asking me when voice; here voice. But it not work good

LordOfPolls committed:
fix: Calm down flake8

LordOfPolls committed:
fix: Improve vc response to being moved

LordOfPolls committed:
fix: Fix errant fix

LordOfPolls committed:
fix: Handle user leaving vc after restart

LordOfPolls committed:
fix: fix logic error in playing prop

LordOfPolls committed:
fix: fix resume and pause

LordOfPolls committed:
refactor: rename VoiceState.py -> active_voice_state.py

LordOfPolls committed:
feat: expose player stopped event

LordOfPolls committed:
fix: Minor tweaks to player logic

start is now set after the first packet has been sent to stop the `fast forward` effect when starting to play

minor logic tweaks

LordOfPolls committed:
feat: Audio buffering

This is to prevent ffmpeg stuttering

LordOfPolls committed:
feat: Active Voice state now handles its own events

LordOfPolls committed:
feat: allow ffmpeg arg passing

LordOfPolls committed:
fix: missing code from previous commit

this should have been included in an earlier commit, mb

LordOfPolls committed:
fix: incorrect event used in player.stopped

LordOfPolls committed:
feat: improve audio buffer (see desc.)

The audio buffer had 2 issues
- The buffer would be empty until it was given some data by the read_ahead task, which the player would interpret as no more audio, and exit
- The buffer could have a bit of data left, if the audio data didnt perfectly divide into the frame length

This fixes both. AudioBuffer no longer waits for a complete frame, instead it pads its output to be the correct length. Note this only occurs at the end of a file
AudioBuffer now has an event which fires when the buffer has been *initialised* (received its first data). The audio object waits for this event when creating processes

LordOfPolls committed:
fix: typo

LordOfPolls committed:
fix: tweak voice reconnect logic

LordOfPolls committed:
build: Add voice extra requirement

LordOfPolls committed:
build: Future proof setup.py with `all` extra

LordOfPolls committed:
build: use setuptools.build_meta

LordOfPolls committed:
ci: pytest to install all deps

LordOfPolls committed:
docs: Use pip instead of requirements.txt

benwoo1110 commented on dis_snek/api/gateway/gateway.py:
```diff
@@ -407,7 +211,7 @@ async def dispatch_event(self, data, seq, event) -> None:
                 self.session_id = data["session_id"]
                 log.info("Connected to gateway!")
-                log.debug(f"Session ID: {self.session_id} Trace: {self._trace}")
+                log.debug(f" Session ID: {self.session_id} Trace: {self._trace}")
```
_random space?_

LordOfPolls replied:
Fixed in 2e7b8741367ffd8559cb5b82e18d853bef53c9f5

benwoo1110 commented on dis_snek/models/discord/channel.py:
```diff
@@ -1734,6 +1735,16 @@ def voice_members(self) -> List["models.Member"]:
         return [self._client.cache.get_member(self._guild_id, member_id) for member_id in self._voice_member_ids]
 
+    @property
+    def voice_state(self) -> "ActiveVoiceState":
```
```suggestion
    def voice_state(self) -> Optional["ActiveVoiceState"]:
```

benwoo1110 commented on dis_snek/client/smart_cache.py:
```diff
@@ -772,6 +774,41 @@ def delete_voice_state(self, user_id: "Snowflake_Type") -> None:
 
+    # region Bot Voice cache
+
+    def get_bot_voice_state(self, guild_id: "Snowflake_Type") -> Optional[ActiveVoiceState]:
```
To standardise with #375, should allow for optional guild_id being passed in

LordOfPolls replied:
Fixed in 6a0c9b4ffe118a4a651a61c4e7fa1e36ddeac9da

LordOfPolls committed:
feat: Add ActiveVoiceState to namespace

LordOfPolls committed:
feat: expose entry in ytdl_audio

LordOfPolls committed:
feat: Ingest guild_creation voice_state data

LordOfPolls committed:
feat: add member.voice

Shows the voice channel a member is currently in

LordOfPolls committed:
feat 💥: member.voice now returns voice state

LordOfPolls committed:
fix: handle opus being destroyed during instantiation

LordOfPolls committed:
build: include dll files in package

LordOfPolls committed:
fix: Fill buffer before starting playback

Deals with ffmpeg struggling to start playback with certain streams

LordOfPolls committed:
feat: Add connected prop

LordOfPolls committed:
fix: Change disconnection logic in active voice state

LordOfPolls committed:
docs: Missing typehints

LordOfPolls committed:
docs: add missing docstrings

LordOfPolls committed:
feat: Option to wait for completion in play method

LordOfPolls committed:
fix: Stop raising RuntimeError on send with no ws

It isn't actually necessary, and only scares end users.

LordOfPolls committed:
docs: ANN204

LordOfPolls committed:
docs: ANN204 cont.

LordOfPolls committed:
docs: ANN204 - I hate merge conflicts

LordOfPolls committed:
feat: Add player.elapsed_time

LordOfPolls committed:
fix: Add dunder all

LordOfPolls committed:
feat: handle discord's poop voice servers

LordOfPolls committed:
fix: Stop error when voice state volume set before playing

LordOfPolls committed:
feat💥: Remove AudioVolume volume init argument

LordOfPolls committed:
fix: Remove volume attr from YTDLAudio

LordOfPolls committed:
feat: Handle stopping while paused

LordOfPolls committed:
fix: Update voice to comply with new code quality req

LordOfPolls committed:
feat: Use context managers for player

LordOfPolls committed:
feat: Add player paused prop

LordOfPolls committed:
feat: Player is now aware if more audio is expected

Bluenix2 commented on dis_snek/api/voice/player.py:
```diff
@@ -0,0 +1,145 @@
+log = logging.getLogger(logger_name)
+
+
+class Player(threading.Thread):
```
Just curious, have you done timing on the voice here? How long does it take to compute (ie. time blocking the event loop)?

LordOfPolls replied:
It wasnt blocking the event loop that was the problem I was solving, it was voice not firing consistently enough when running as an asyncio task. Not to mention, with discord's voice implimentation, you **must** send the payload on time, or you will experience stuttering. As such a threaded solution is prefered over an asyncio task. 

If voice sends too early, you get fast-forward artifacts, if it fires too late, you get stuttering or discord will just outright ignore the packet

LordOfPolls committed:
fix: Audio read returning bytearray not bytes

LordOfPolls committed:
fix: Paused was always opposite

LordOfPolls committed:
fix: VoiceState connected set incorrectly

LordOfPolls committed:
feat: Add play_no_wait

LordOfPolls committed:
fix: Handle player not existing in voicestate props

LordOfPolls committed:
docs: Voice docs

zevaryx reviewed (approved):
LGTM, have been testing it as changes have been made, and have been following it closely

LordOfPolls committed:
Style: Fix errant space

LordOfPolls committed:
docs: Fix type hinting of voice_state

Co-authored-by: Ben Woo <30431861+benwoo1110@users.noreply.github.com>

LordOfPolls committed:
feat: Allow optional guild_id for get_bot_voice_state

benwoo1110 committed:
fix: Voice gateway resume payload

benwoo1110 committed:
fix: Don't move bot if already in that voice channel

benwoo1110 committed:
fix: Various issues with voice disconnection

* Only stop player if playing
* Fix logic error where disconnect method is ran twice
* Fix issue with websocket not closing properly

benwoo1110 committed:
feat: Add support for self_mute and self_deaf

benwoo1110 committed:
perf: Improve snowflake comparison

Int compare is faster than str

benwoo1110 committed:
fix: Private _on_raw_voice_server_update method

benwoo1110 committed:
fix: Handle exception during voice connection

* Raise exception if voice already connected
* Raise exception if voice event isnt received

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

benwoo1110 committed:
fix: Add voice exception classes to `__all__`

LordOfPolls committed:
fix: r100- raise without from

benwoo1110 committed:
fix: `play_no_wait` doesn't need to be async

LordOfPolls committed:
fix: remove redundant player instantiation

benwoo1110 committed:
docs: Add missing docstring for voice related methods

LordOfPolls committed:
fix: Fix AVoiceState move_to

- Allow connection calls if moving to a new channel
- Pause player pre-emptively, before re-creating voice gateway connection

LordOfPolls committed:
fix: Further improve movement handling

- Rather than re-creating the voice websocket, simply send a voice state update, and handle it using the exact same methods as that would be ran if a user manually moved the bot

LordOfPolls committed:
fix: account for player already being paused before move

benwoo1110 committed:
docs: Fix filename typo

benwoo1110 reviewed (approved):
Tested and fixed all the bugs i found

Kigstn commented on dis_snek/models/discord/user.py:
```diff
@@ -348,6 +349,11 @@ def guild_permissions(self) -> Permissions:
         return permissions
 
+    @property
+    def voice(self) -> Optional["VoiceState"]:
```
This is inconsistent with other usages

Kigstn replied:
```suggestion
    def voice_state(self) -> Optional["VoiceState"]:
```

Kigstn commented on dis_snek/client/client.py:
```diff
@@ -1831,6 +1832,27 @@ async def fetch_voice_regions(self) -> List["VoiceRegion"]:
+        """
+        return await self._connection_state.voice_connect(guild_id, channel_id, muted, deafened)
+
+    def get_bot_voice_state(self, guild_id: "Snowflake_Type") -> ActiveVoiceState:
```
This is missing a docstring.

Kigstn replied:
This should also have a helper method in `Guild`

Kigstn replied:
Wrong return annotation
```suggestion
    def get_bot_voice_state(self, guild_id: "Snowflake_Type") -> Optional[ActiveVoiceState]:
```

Kigstn commented on docs/src/Guides/22 Voice.md:
```diff
@@ -0,0 +1,34 @@
+Now just join a voice channel, and type run the "play" slash command with a song of your choice.
+
+Congratulations! You've got a music-bot.
+
```
This should include a section on how to play your own files, like this f.e.

```suggestion

## Your Own Tunes

If you want to play your own audio, you can do that as well. Just create an `AudioVolume` object and that's it!


```py
audio = await AudioVolume("my_audio_file.mp3")
await ctx.voice_state.play(audio)
```

```

Kigstn commented on dis_snek/api/voice/audio.py:
```diff
@@ -0,0 +1,251 @@
+        ...
+
+
+class Audio(BaseAudio):
```
IMO this needs to supports for raw bytes to be input

LordOfPolls replied:
Thats a niche use-case, that imo should be handled by the user themselves, not the library. Namely because we wont know anything about the bytestream.

Kigstn commented on dis_snek/models/discord/guild.py:
```diff
@@ -216,6 +216,12 @@ def _process_dict(cls, data: Dict[str, Any], client: "Snake") -> Dict[str, Any]:
 
         if welcome_screen := data.get("welcome_screen"):
             data["welcome_screen"] = GuildWelcome.from_dict(welcome_screen, client)
+
```
This should be remade into a property that returns dict[user_id, VoiceState]. Currently01, `Guild.voice_states` just returns the raw dict

![image](https://user-images.githubusercontent.com/38250010/160820907-ec887943-ef56-482f-94a4-d018a6b4dcd6.png)

LordOfPolls replied:
Thats not the fault of this code snippet. Caching a voice state is async, so it made more sense to do it in tasks to prevent bogging down when caching *busy* guilds. 

I agree a property should be made, adding it now

LordOfPolls replied:
Resolved in a nice ick way. 

8baddbf3a39c4bd1a0dbd4da3bd4eae5d79ebfb3

Kigstn commented on docs/src/Guides/22 Voice.md:
````diff
@@ -0,0 +1,34 @@
+```
+pip install dis-snek[voice]
+```
+Now you've got those; let's make a simple play command to get you started.
````
Inform the users they need to have FFmpeg installed

```suggestion
Also, make sure you have installed [FFmpeg](https://ffmpeg.org/) and made it available in the root directory.

Now you've got those; let's make a simple play command to get you started.
```

benwoo1110 committed:
fix: `get_bot_voice_state` optional type hinting

Co-authored-by: Daniel J <38250010+Kigstn@users.noreply.github.com>

benwoo1110 committed:
feat: Add `Guild.voice_state` property

benwoo1110 committed:
docs: Add docstring for `Snake.get_bot_voice_state`

benwoo1110 committed:
docs: Add note of downloading FFmpeg

Co-authored-by: Daniel J <38250010+Kigstn@users.noreply.github.com>

LordOfPolls committed:
feat: Add check for ffmpeg during player init

LordOfPolls committed:
docs: Cover playing local files

LordOfPolls committed:
docs: Rewrite ffmpeg note

You don't install ffmpeg on a lot of systems, you just download it. Also the previous note didnt consider PATH which is a common location to "install" it

LordOfPolls committed:
style: Fix capitalisation of FFmpeg in runtime error

LordOfPolls committed:
feat: guild.voice_states returns objects, not dicts
